Add Commit::extract_signature

carlosmn opened the pull request:
This returns the signature and the signed data so the user can compare
them.

---

This depends on libgit2/libgit2#3599 so we shouldn't merge until that gets resolved.

arthurschreiber commented on ext/rugged/rugged_commit.c:
```diff
@@ -612,11 +612,53 @@ static VALUE rb_git_commit_header(VALUE self)
+
+	rugged_exception_check(error);
+
+	ret_arr = rb_ary_new3(2, rb_enc_str_new(signature.ptr, signature.size, rb_utf8_encoding()),
```
Shouldn't this use `git_commit_message_encoding` instead of hardcoding UTF-8?

vmg replied:
actually, since we're meant to verify these bytes, they'd be better off as a binary/ascii string instead.

carlosmn replied:
Yeah, they're really just bytes. Is there a better type to use for this? The message encoding is only really supposed to be about the commit message, even if we end up using it for quite a few more things.

vmg replied:
Yeah, tag it as `ASCII_8BIT`, that's the default "binary" encoding in Ruby.

carlosmn replied:
Changed the encoding to `rb_ascii8bit_encoding()` which should be the right one. We don't seem to use it anywhere else in the bindings, which seems odd, though;

vmg replied:
Hm. We should be using it. Let me double check.

carlosmn replied:
For the blob we use `rb_str_new()` which seems to do the same thing with fewer chars, so maybe use that instead of specifying it?

vmg replied:
Right. That's what I was missing. Let's go with `rb_str_new`.

brianmario commented:
@carlosmn I bumped libgit2 to pull in the new call you added, things appear to be working as expected. Was there anything else you wanted to get in here before merging? Possibly an instance method on `Rugged::Commit` for this that wraps this one but with a single [optional] `field` string or something?

arthurschreiber commented:
@carlosmn The commit (and PR title) refer to `extract_header`, while the method is called `extract_signature`. I guess the latter is correct?

carlosmn commented:
> The commit (and PR title) refer to extract_header, while the method is called extract_signature. I guess the latter is correct?

Yeah, I often think I could do with copy editor :)

>Possibly an instance method on Rugged::Commit for this that wraps this one but with a single [optional] field string or something?

We could do with an instance method, but it'd just be `def extract_signature; Commit::extract_signature(self.hex); end` so I'm a bit meh. I do still want to add  a `field` param to the static method (class method?) which was a change made after I wrote this version.

carlosmn commented:
Changed the commit message, added the optional signature argument and rebased so the order makes sense squashing these together. Should be ready for merging now.
